// mp-colorui/dist/h5/components/switch/h5/index.js
import Nerv from "nervjs";
import { View } from "@tarojs/components";
import Taro from "@tarojs/taro-h5";
export default class ClSwitch_h5 extends Taro.Component {
  onChange(flag) {
    this.props.onChange && this.props.onChange(flag);
  }
  render() {
    const title = this.props.title;
    const color = this.props.color || 'green';
    const shapeClassName = this.props.shape !== 'radius' ? '' : 'radius';
    const type = this.props.type === 'form' ? 'form' : 'normal';
    const checked = !!this.props.checked;
    const [checkedSwitch, setCheckedSwitch] = Taro.useState(checked);
    const switchComponent = <View className={`h5-switch ${checkedSwitch ? 'checked' : ''} ${color}`} onClick={() => {
      const currentCheck = !checkedSwitch;
      setCheckedSwitch(currentCheck);
      this.onChange(currentCheck);
    }}>
        <View className={`h5-switch-input sm ${shapeClassName}`} />
      </View>;
    const formSwitchComponent = <View className="cu-form-group">
        <View className="title">{title}</View>
        {switchComponent}
      </View>;
    return type === 'form' ? formSwitchComponent : switchComponent;
  }
}
ClSwitch_h5.options = {
  addGlobalClass: true
};
ClSwitch_h5.defaultProps = {
  title: '',
  color: 'green',
  shape: 'normal',
  type: 'normal',
  checked: false,
  onChange: () => {}
};
