// mp-colorui/dist/h5/components/accordion/index.js
import Nerv from "nervjs";
import Taro, { createSelectorQuery as _createSelectorQuery } from "@tarojs/taro-h5";
import { View } from '@tarojs/components';
import ClIcon from "../icon/index";
export default class ClAccordion extends Taro.Component {
  render() {
    const props = this.props;

    const { open, title, animation, onClick, card } = props;
    const [show, setShow] = Taro.useState(open);
    const [height, setHeight] = Taro.useState(0);
    const id = `cl-accordion-content-${+new Date()}`;
    Taro.useEffect(() => {
      if (show) {
        const query = _createSelectorQuery().in(this.$scope);
        query.select(`#${id}`).fields({
          size: true,
          rect: true
        }, res => {
          setHeight(res.height);
        }).exec();
      } else {
        setHeight(0);
      }
    }, [show]);
    return <View className={`cu-list menu ${card ? 'card-menu' : ''}`} style={{
      overflow: "hidden",
      height: Taro.pxTransform((50 + height) * 2),
      transition: `all ${animation ? `${props.speed}s` : 0} linear`,
      color: 'black'
    }}>
      <View className={`cu-item`} onClick={() => {
        onClick && onClick(!show);
        setShow(!show);
      }}>
        <View className="flex justify-between align-center" style={{ width: '100%' }}>
          <View>{title}</View>
          <View style={{ transition: `all ${props.speed}s linear`, transform: `rotate(${show ? '90deg' : '0'})` }}>
            <ClIcon iconName={"right"} size={"small"} />
          </View>
        </View>
      </View>
      <View id={id} className="bg-white">
        {this.props.children}
      </View>
    </View>;
  }

}
ClAccordion.options = {
  addGlobalClass: true
};
ClAccordion.defaultProps = {
  open: false,
  title: '',
  animation: true,
  onClick: () => {},
  card: false,
  speed: 0.15
};
